src/Microsoft.Health.Dicom.SqlServer/Features/Schema/Migrations/4.diff.sql

```diff
@@ -151,6 +134,50 @@ AS
           AND Status = @status
 GO
 
+/***************************************************************************************/
+-- STORED PROCEDURE
+--     GetInstanceBatches
+--
+-- DESCRIPTION
+--     Divides up the instances into a configurable number of batches.
+--
+-- PARAMETERS
+--     @batchSize
+--         * The desired number of instances per batch. Actual number may be smaller.
+--     @batchCount
+--         * The desired number of batches. Actual number may be smaller.
+--     @maxWatermark
+--         * The optional exclusive maximum watermark.
+--
+-- RETURN VALUE
+--     The batches as defined by their inclusive minimum and maximum values.
+/***************************************************************************************/
+CREATE OR ALTER PROCEDURE dbo.GetInstanceBatches (
+    @batchSize INT,
+    @batchCount INT,
+    @maxWatermark BIGINT = NULL
+)
+AS
+BEGIN
+    SET NOCOUNT     ON
+    SET XACT_ABORT  ON
+
+    SELECT
+        MIN(Watermark) AS MinWatermark,
+        MAX(Watermark) AS MaxWatermark
+    FROM
+    (
+        SELECT TOP (@batchSize * @batchCount)
+            Watermark,
+            (ROW_NUMBER() OVER(ORDER BY Watermark DESC) - 1) / @batchSize AS Batch
+        FROM dbo.Instance
+        WHERE @maxWatermark IS NULL or Watermark < @maxWatermark
```

Comment:
Do we want to make @maxWatermark inclusive,  since everywhere is Inclusive?
Or add comments here explicit tells @maxWatermark is exclusive?

Reply:
nit: or =>OR

Reply:
Here, it's a bit different, as we often pass in the last "start" value which we've already processed. So we could instead pass something like "start - 1" which would be fine too. It would also make the nullable value easier to process because would could do something like `Watermark <= ISNULL(@maxWatermark, Watermark)`